### PR TITLE
[SPARK-28266][SQL] convertToLogicalRelation should not interpret `path` property when reading Hive tables

### PR DESCRIPTION
**Pull request by shardulm94**
### What changes were proposed in this pull request?

For non-datasource Hive tables, e.g. tables written outside of Spark (through Hive or Trino), we have certain optimzations in Spark where we use Spark ORC and Parquet datasources to read these tables ([Ref](https://github.com/apache/spark/blob/fbf53dee37129a493a4e5d5a007625b35f44fbda/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala#L128)) rather than using the Hive serde.
If such a table contains a `path` property, Spark will try to list this path property in addition to the table location when creating an `InMemoryFileIndex`. ([Ref](https://github.com/apache/spark/blob/fbf53dee37129a493a4e5d5a007625b35f44fbda/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L575)) This can lead to wrong data if `path` property points to a directory location or an error if `path` is not a location. A concrete example is provided in [SPARK-28266 (comment)](https://issues.apache.org/jira/browse/SPARK-28266?focusedCommentId=17380170&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17380170).

Since these tables were not written through Spark, Spark should not interpret this `path` property as it can be set by an external system with a different meaning.

### Why are the changes needed?

For better compatibility with Hive tables generated by other platforms (non-Spark)

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit test

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 1]
```diff
@@ -363,4 +363,33 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
+  Seq(
+    "parquet" -> "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
+    "orc" -> "org.apache.hadoop.hive.ql.io.orc.OrcSerde"
+  ).foreach { case (format, serde) =>
+    test("SPARK-28266: convertToLogicalRelation should not interpret `path` property when " +
+      s"reading Hive tables using $format file format") {
+      withTempPath(dir => {
+        val baseDir = s"${dir.getCanonicalFile.toURI.toString}"
+        spark.range(3).selectExpr("id").write.format(format).save(baseDir)
+        withTable("non_partition_table") {
```
Comment (xkrogen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala:376]: Can we give the table a more simple name like `foo` or `t`? Or is it relevant that it's not partitioned?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 2]
```diff
@@ -363,4 +363,33 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
   }
+
+  Seq(
+    "parquet" -> "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
+    "orc" -> "org.apache.hadoop.hive.ql.io.orc.OrcSerde"
+  ).foreach { case (format, serde) =>
+    test("SPARK-28266: convertToLogicalRelation should not interpret `path` property when " +
+      s"reading Hive tables using $format file format") {
+      withTempPath(dir => {
+        val baseDir = s"${dir.getCanonicalFile.toURI.toString}"
```
Comment (xkrogen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala:374]: Seems that the string interpolation is unnecessary here?

Also can we just use `dir.getAbsolutePath`? Or do we require a URI here?

Comment (shardulm94) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Fixed! Copied it from a test above, didn't pay too much attention.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala (hunk 1 above)
Comment (shardulm94) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Used simple names now. It is somewhat relevant that the table is non-partitioned. For partitioned tables, the location of each partition is used and so these table level properties are ignored.

**Comment by xkrogen** [comment]
cc @dongjoon-hyun @cloud-fan @gatorsmile @viirya

**Review by viirya — COMMENTED** [review]
Hmm, are we sure `path` is not used at all for reading files from the table from all external systems? This change seems to break some cases if `path` is actually used?

**Comment by shardulm94** [comment]
@viirya I cannot comment about ALL external systems, but at least the ones that I am familiar with (Hive/Trino/Presto/Iceberg) do not use this property when writing tables. It is possible that some external system may be using this property, but even in that case, we seem to be using BOTH the table location and the property to prepare the list of files, thus double counting the files. In any case, respecting the `path` property for such tables does not seem to be intentional at all.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 3]
```diff
@@ -244,7 +244,11 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
                 paths = rootPath.toString :: Nil,
                 userSpecifiedSchema = Option(updatedTable.dataSchema),
                 bucketSpec = None,
-                options = options,
+                // Do not interpret the 'path' option at all when tables are read using the Hive
+                // source, since the URIs will already have been read from the table's LOCATION.
+                // The `Map() ++` is necessary to force materialization since the return of
+                // `filterKeys` is a view which is not serializable
+                options = Map() ++ options.filterKeys(!_.equalsIgnoreCase("path")),
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:251]: I agree, `path` is a special data source option in Spark. For hive tables not written by Spark, it's wrong to apply the Spark knowledge here and treat the `path` option specially.

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:251]: nit: `options.filterKeys(!_.equalsIgnoreCase("path")).toMap`?

Comment (xkrogen) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:251]: Unfortunately `toMap` is insufficient to materialize the view, check out the Scaladoc/implementation:
```
  /** Returns this $coll as an immutable map.
   *
   *  A new map will not be built; lazy collections will stay lazy.
   */
  @deprecatedOverriding("Immutable maps should do nothing on toMap except return themselves cast as a map.",  "2.11.0")
  override def toMap[T, U](implicit ev: (K, V) <:< (T, U)): immutable.Map[T, U] =
    self.asInstanceOf[immutable.Map[T, U]]
```

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:251]: how about `options.filter { case (k, _) => !k.equalsIgnoreCase("path") }`?

Comment (shardulm94) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Done

**Comment by cloud-fan** [comment]
ok to test

File: .idea/vcs.xml [hunk 4]
```diff
@@ -1,24 +1,16 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
```
Comment (cloud-fan) [inline .idea/vcs.xml:2]: unnecessary change?

**Review by cloud-fan — APPROVED** [review]

File: .idea/vcs.xml (hunk 4 above)
Comment (xkrogen) [reply .idea/vcs.xml:2]: btw @shardulm94 you can do this to tell `git` to ignore local changes to this file:
```
git update-index --skip-worktree .idea/vcs.xml
```

Comment (shardulm94) [reply .idea/vcs.xml]: My bad! IntelliJ modified the file and I didn't realize when pushing an empty commit. @xkrogen Thanks for the tip!

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2/3.1/3.0!

**Comment by yaooqinn** [comment]
Late LGTM.

**Comment by gatorsmile** [comment]
@shardulm94 We should add the behavior change to the migration guide, although this a bug fix.
